src/main/groovy/org/gradle/android/workarounds/room/KspWorkaround.groovy

```diff
@@ -0,0 +1,93 @@
+            configureWorkaroundTask(it)
+        }
+    }
+
+    @Override
+    void configureWorkaroundTask(Task task) {
+
+        def fileOperations = project.fileOperations
+
+        addArgumentProvider(task)
+
+        task.doFirst onlyIfKspRoomSchemaLocationArgumentProviderIsConfigured(task.commandLineArgumentProviders) { KspRoomSchemaLocationArgumentProvider provider ->
+            KspWorkaround.copyExistingSchemasToTaskSpecificTmpDir(fileOperations, roomExtension.schemaLocationDir, provider)
+        }
+
+        task.doLast onlyIfKspRoomSchemaLocationArgumentProviderIsConfigured(task.commandLineArgumentProviders) { KspRoomSchemaLocationArgumentProvider provider ->
+            KspWorkaround.copyGeneratedSchemasToOutput(fileOperations, provider)
+        }
+
+        task.finalizedBy onlyIfKspRoomSchemaLocationArgumentProviderIsConfigured(task.commandLineArgumentProviders) {
+            roomExtension.schemaLocationDir.isPresent() ? mergeTask : null
+        }
+
+        TaskExecutionGraph taskGraph = project.gradle.taskGraph
+        taskGraph.whenReady onlyIfKspRoomSchemaLocationArgumentProviderIsConfigured(task.commandLineArgumentProviders) { KspRoomSchemaLocationArgumentProvider provider ->
+            if (taskGraph.hasTask(task)) {
+                roomExtension.registerOutputDirectory(provider.schemaLocationDir)
+            }
+        }
+    }
+
+    private void addArgumentProvider(Task task) {
+        if (roomExtension.schemaLocationDir.isPresent()) {
+            def variantSpecificSchemaDir = project.objects.directoryProperty()
+            variantSpecificSchemaDir.set(androidVariantProvider.getVariantSpecificSchemaDir(project, "${task.name}"))
+            task.commandLineArgumentProviders.add(new KspRoomSchemaLocationArgumentProvider(roomExtension.schemaLocationDir, variantSpecificSchemaDir))
+        }
+    }
+
+    static void copyExistingSchemasToTaskSpecificTmpDir(FileOperations fileOperations, Provider<Directory> existingSchemaDir, KspRoomSchemaLocationArgumentProvider provider) {
+        if (existingSchemaDir.isPresent()) {
+            def temporaryVariantSpecificSchemaDir = provider.temporarySchemaLocationDir
+            fileOperations.sync {
+                it.from existingSchemaDir
+                it.into temporaryVariantSpecificSchemaDir
+            }
+        }
+    }
+
+    static void copyGeneratedSchemasToOutput(FileOperations fileOperations, KspRoomSchemaLocationArgumentProvider provider) {
+        def variantSpecificSchemaDir = provider.schemaLocationDir
+        def temporaryVariantSpecificSchemaDir = provider.temporarySchemaLocationDir
+        fileOperations.sync {
+            it.from temporaryVariantSpecificSchemaDir
+            it.into variantSpecificSchemaDir
+        }
+    }
+
+    private static Closure onlyIfKspRoomSchemaLocationArgumentProviderIsConfigured(def commandLineArgumentProviders, Closure<?> action) {
```

Comment:
what is the situation in which we would not add the argument provider when calling this function? is it only `!roomExtension.schemaLocationDir.isPresent()` or is there another case?

Reply:
Ksp can be applied for other processors like moshi, following your comments, we just apply the workaround if the extension exists and removed the redundant verification of `onlyIfKspRoomSchemaLocationArgumentProviderIsConfigured`

https://github.com/gradle/android-cache-fix-gradle-plugin/pull/429/commits/bfd183a64212dc15ce7b58756d790e0fec0f8568